### PR TITLE
preserve class type when using from_queue

### PR DESCRIPTION
**Pull request by timmysilv**
**Context:**
`QuantumScript.from_queue(q)` would always return a `QuantumScript`, even if a user calls it from a child class. This would be confusing/unexpected.

**Description of the Change:**
Return a script with a type matching the calling class when doing `QuantumScript_or_child_class.from_queue(q)`.

**Benefits:**
Less confusion. I am mainly adding this to enable users to do `QuantumTape.from_queue`. Before, it would return a `QuantumScript`. With this PR, it will return a `QuantumTape`. If all goes to plan, we will change `QuantumTape` to become what is today known as `QuantumScript`, so this utility will allow for a consistent API for those users who do not wish to leave `QuantumTape` in the interim.

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**
N/A

### CODE REVIEW

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -300,6 +300,9 @@
   [(#3399)](https://github.com/PennyLaneAI/pennylane/pull/3399)
 
 * Improved the performance of executing circuits under the `jax.vmap` transformation, which can now leverage the batch-execution capabilities of some devices. [(#3452)](https://github.com/PennyLaneAI/pennylane/pull/3452)
+
+* Child classes of `QuantumScript` now return their own type when using `SomeChildClass.from_queue`.
+  [(#3502)](https://github.com/PennyLaneAI/pennylane/pull/3502)
```
Comment (timmysilv) [inline doc/releases/changelog-dev.md:305]:
```suggestion
  [(#3501)](https://github.com/PennyLaneAI/pennylane/pull/3501)
```

**Review by rmoyard — COMMENTED** [review]
It looks good to me, thanks 👍  did it appear in the OperationRecorder PR? You can add a small bug changelog item.

**Comment by timmysilv** [comment]
I'll add something to the bug section, but it didn't really come from a bug concern. @josh146 and I were chatting about the future for class naming, and we will probably rename what is today `QuantumScript` back to `QuantumTape`, and delete what is today called `QuantumTape`. So - this allows people to do `QuantumTape.from_queue(q)`, receive a `QuantumTape`, and when we make the switch, the difference will be imperceptible!

File: pennylane/tape/qscript.py [hunk 2]
```diff
@@ -1293,10 +1293,10 @@ def to_openqasm(self, wires=None, rotations=True, measure_all=True, precision=No
 
         return qasm_str
 
-    @staticmethod
-    def from_queue(queue):
+    @classmethod
+    def from_queue(cls, queue):
         """Construct a QuantumScript from an AnnotatedQueue."""
-        return QuantumScript(*process_queue(queue))
+        return cls(*process_queue(queue))
```
Comment (Jaybsoni) [inline pennylane/tape/qscript.py:1299]: What if the init for the sub class is different, would this break?

Comment (timmysilv) [reply pennylane/tape/qscript.py:1299]: yes it would! adding a test to block this rn

**Comment by timmysilv** [comment]
my latest commit added two tests:
1. A test introducing a subclass of QuantumScript with a different __init__, and asserts that from_queue fails for that class
2. A test that iterates through _actual current_ subclasses of QuantumScript, and asserts that they all succeed in calling `from_queue`

File: pennylane/tape/qscript.py (hunk 2 above)
Comment (timmysilv) [reply pennylane/tape/qscript.py:1299]: see latest comment for details

**Pull request merged**